core/templates/dev/head/services/PlaythroughServiceSpec.ts

```diff
@@ -18,11 +18,19 @@
 
 require('App.ts');
 require('domain/statistics/LearnerActionObjectFactory.ts');
-require('services/ExplorationFeaturesService.ts');
 require('services/PlaythroughService.ts');
 
 describe('PlaythroughService', function() {
   beforeEach(angular.mock.module('oppia'));
+  beforeEach(angular.mock.module(function($provide) {
```

Comment:
Why are we mocking rather than using the actual service?

If done to avoid dependency complications, are there any examples of unit tests in Angular 8 that _don't_ mock? I'm trying to understand what constraints we should be operating under in the future when writing unit tests, and not being able to import anything else seems very restrictive.

Reply:
If I don't mock I'll get `Uncaught Error: Trying to get the Angular injector before bootstrapping an Angular module.`. This is because the file is AngularJS and the dependency in Angular. Angular is not bootstrapped. In real fully Angular tests, there is something called `Testbed` that is an internal testing utility of Angular that can provide dependencies.

Reply:
Right -- part of doing this for a single service first is to resolve all those complications, so that we have a pattern that can be extended to other tests. So if our eventual Karma tests would allow importing of modules, we should solve that in this PR.

Reply:
I don't think I understand. The importing of files in Karma is supposed to happen when we use Testbed - which requires both the file and the corresponding dependency to be injected to be an Angular service (since Testbed is an Angular utility).

Are you trying to say that I try to upgrade multiple services which are in a kind of tree that would test this hypothesis?

Reply:
Sorry, I'll rephrase. Let's say some test needs to use a dependency and that dependency is too complicated to mock (or we actually want to test the full flow across multiple files).

(a) How would you handle this case, during the upcoming migration, if the dependency is still AngularJS and the test is in an Angular file?

(b) How would you handle this case in the long term?

Reply:
As it stands now, since the Angular is loaded lazily on the encounter of the first Angular component and I can't seem to find a solution to this. So, the dependencies still need to be mocked. And this needs to be done for all the dependencies that are used in the Spec. 
So (a) -> The AngularJS dependencies would have to be made as mocks in Angular tests. There is very much ambiguity and uncertainty in testing hybrid apps. The way Angular team has done it is by making a mini-module which has just the things needed for the test and such that they are isolated. (This was outlined by one of the members of Angular team [here](https://github.com/angular/angular/issues/24369#issuecomment-395843749)).

(b) -> For the long term, this has to be done as long as the dependencies and the main file are from different versions. Once a dependency of an Angular service becomes Angular, it can be injected using the Angular framework.

Reply:
OK, thanks for the explanation. I'm fine with you doing mocks where it makes sense. You probably need a plan for the larger dependencies though, maybe you can migrate the final chunk all at once (as long as it's not too big).

Reply:
Yes, that is actually the plan, migrating the big ones at the last.